### PR TITLE
Fix EffectRenderer keeping reference to world instance when leaving world

### PR DESCRIPTION
**Pull request by tth05**

### CODE REVIEW

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinMinecraft_ClearEffectRenderer.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Shadow
+    public EffectRenderer effectRenderer;
+
+    @Inject(
+            method = "loadWorld(Lnet/minecraft/client/multiplayer/WorldClient;Ljava/lang/String;)V",
+            at = @At(value = "HEAD"))
+    private void hodgepodge$fixEffectRendererWorldLeak(WorldClient worldClient, String loadingMessage,
+            CallbackInfo ci) {
+        if (worldClient == null) {
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinMinecraft_ClearEffectRenderer.java:24]: I see that the vanilla code in this method is using a null check on the effectRenderer field before calling that method, can you do the same thing ?

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinMinecraft_ClearEffectRenderer.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+@Mixin(Minecraft.class)
+public abstract class MixinMinecraft_ClearEffectRenderer {
+
+    @Shadow
+    public EffectRenderer effectRenderer;
+
+    @Inject(
+            method = "loadWorld(Lnet/minecraft/client/multiplayer/WorldClient;Ljava/lang/String;)V",
+            at = @At(value = "HEAD"))
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinMinecraft_ClearEffectRenderer.java:21]: can you inject right before it assigns the `this.theWorld = worldClientIn;` and not at the head of the method please

**Review by Alexdoru — CHANGES_REQUESTED** [review]

**Comment by Alexdoru** [comment]
This method says to call it with null to clean, why not do it as well, could that be a leak too

image

**Comment by tth05** [comment]
I also added another fix for RenderGlobal

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinMinecraft_ClearRenderersWorldLeak.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Shadow
+    public RenderGlobal renderGlobal;
+
+    @Inject(
+            method = "loadWorld(Lnet/minecraft/client/multiplayer/WorldClient;Ljava/lang/String;)V",
+            at = @At(
+                    value = "FIELD",
+                    target = "Lnet/minecraft/client/Minecraft;theWorld:Lnet/minecraft/client/multiplayer/WorldClient;",
+                    opcode = Opcodes.PUTFIELD))
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinMinecraft_ClearRenderersWorldLeak.java:29]: can you put `shift = At.Shift.AFTER`

**Review by Alexdoru — CHANGES_REQUESTED** [review]

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinRenderGlobal_FixWordLeak.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    @Inject(
+            method = "setWorldAndLoadRenderers(Lnet/minecraft/client/multiplayer/WorldClient;)V",
+            at = @At(
+                    value = "FIELD",
+                    target = "Lnet/minecraft/client/renderer/RenderGlobal;theWorld:Lnet/minecraft/client/multiplayer/WorldClient;",
+                    opcode = Opcodes.PUTFIELD))
+    public void hodgepodge$clearWorldRenderersAndTileEntities(WorldClient worldClient, CallbackInfo ci) {
+        if (worldClient == null) {
+            tileEntities.clear();
+            worldRenderers = null;
```
[the remark this answers is not in the document]
Comment (tth05) [reply src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinRenderGlobal_FixWordLeak.java:37]: Sure, although that method does not do anything useful

[the remark this answers is not in the document]
Comment (Alexdoru) [reply src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinRenderGlobal_FixWordLeak.java:45]: it sets the world field to null, idk if that's useful, i'm suggesting

[the remark this answers is not in the document]
Comment (Alexdoru) [reply src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinRenderGlobal_FixWordLeak.java:45]: if you think it is useless you can remove it

**Review by Alexdoru — APPROVED** [review]

**Comment by Alexdoru** [comment]
can you rebase your branch on top of master @tth05

**Pull request merged**